pkg/osquery/table/program_icons_windows.go

```diff
@@ -0,0 +1,195 @@
+				"name":    name,
+				"version": version,
+			})
+		}
+	}
+	return uninstallerIcons
+}
+
+func generateInstallersProgramIcons() []map[string]string {
+	var installerIcons []map[string]string
+
+	productRegPaths := map[registry.Key][]string{
+		registry.CLASSES_ROOT: expandRegistryKey(registry.CLASSES_ROOT, `Installer\Products\*`),
+		registry.USERS:        expandRegistryKey(registry.USERS, `*\Software\Microsoft\Installer\Products\*`),
+	}
+
+	for key, paths := range productRegPaths {
+		for _, path := range paths {
+			key, err := registry.OpenKey(key, path, registry.ALL_ACCESS)
+			defer key.Close()
+			if err != nil {
+				continue
+			}
+
+			iconPath, _, err := key.GetStringValue("ProductIcon")
+			icon, err := parseIcoFile(iconPath)
+			if err != nil {
+				continue
+			}
+			name, _, _ := key.GetStringValue("ProductName")
+			if err != nil {
+				continue
+			}
+
+			installerIcons = append(installerIcons, map[string]string{
+				"icon": icon.base64,
+				"hash": fmt.Sprintf("%x", icon.hash),
+				"name": name,
+			})
+		}
+	}
+
+	return installerIcons
+}
+
+/*
+	Parse an ico returing a base64 encoded version and a hash of the ico.
+
+	This doesn't support extracting an icon from a exe. Windows stores some icon in
+	the exe like 'OneDriveSetup.exe,-101'
+*/
+func parseIcoFile(fullPath string) (icon, error) {
+	var programIcon icon
+	expandedPath, err := registry.ExpandString(fullPath)
+	icoReader, err := os.Open(expandedPath)
+	if err != nil {
+		return programIcon, err
+	}
+	img, err := ico.Decode(icoReader)
+	if err != nil {
+		return programIcon, err
+	}
+	buf := new(bytes.Buffer)
+	img = resize.Resize(128, 128, img, resize.Bilinear)
+	if err := png.Encode(buf, img); err != nil {
+		return programIcon, err
+	}
+
+	checksum := crc64.Checksum(buf.Bytes(), crcTable)
+	return icon{base64: base64.StdEncoding.EncodeToString(buf.Bytes()), hash: checksum}, nil
+}
+
+/*
+	expand registry keys one layer deep for a hive and path.
+
+	For example expandRegistryKey(registry.USERS, `*\Software\Microsoft\Installer\Products\*`)
+	expands to
+	USER1\Software\Microsoft\Installer\Products\2CCC92FB8B3D5F6499511F546A784ACD
+	USER1\Software\Microsoft\Installer\Products\1AAA2FB8B3D5F6499511F546A784ACD
+	USER2\Software\Microsoft\Installer\Products\3FFF92FB8B3D5F6499511F546A784ACD
+	USER2\Software\Microsoft\Installer\Products\5DDD92FB8B3D5F6499511F546A784ACD
+*/
+func expandRegistryKey(hive registry.Key, pattern string) []string {
+	var paths []string
+	magicChar := `*`
+
+	patternsQueue := []string{pattern}
+	for len(patternsQueue) > 0 {
+		expandablePattern := patternsQueue[0]
+		patternsQueue = patternsQueue[1:]
+
+		// add path to results if it doesn't contain the magic char
+		if !strings.Contains(expandablePattern, magicChar) {
+			paths = append(paths, expandablePattern)
+			continue
+		}
+
+		patternParts := strings.SplitN(expandablePattern, magicChar, 2)
+		key, err := registry.OpenKey(hive, patternParts[0], registry.ALL_ACCESS)
```

Comment:
```suggestion
		key, err := registry.OpenKey(hive, patternParts[0], registry.READ)
```

do we need that much permission?